scarb/src/bin/scarb/commands/tree.rs

```diff
@@ -0,0 +1,313 @@
+use crate::args::TreeCommandArgs;
+use anyhow::Result;
+use scarb::core::{Config, Package, PackageId, TargetKind};
+use scarb::ops;
+use scarb_ui::Message;
+use serde::{Serialize, Serializer};
+use std::collections::HashSet;
+use std::fmt::Write;
+
+#[tracing::instrument(skip_all, level = "info")]
+pub fn run(args: TreeCommandArgs, config: &Config) -> Result<()> {
+    let ws = ops::read_workspace(config.manifest_path(), config)?;
+    let packages = args
+        .packages_filter
+        .match_many(&ws)?
+        .into_iter()
+        .map(|p| p.id)
+        .collect::<Vec<_>>();
+
+    let resolve = ops::resolve_workspace(&ws)?;
+
+    let forest = build(packages, &resolve, &args)?;
+
+    config.ui().force_print(forest);
+
+    Ok(())
+}
+
+fn build(
+    packages: Vec<PackageId>,
+    resolve: &ops::WorkspaceResolve,
+    args: &TreeCommandArgs,
+) -> Result<Tree> {
+    struct Visitor<'a> {
+        main_package_id: PackageId,
+        // Tracks visited packages to avoid duplicates among separate branches.
+        visited: &'a mut HashSet<PackageId>,
+        // Tracks visited packages to avoid cycles within a branch.
+        stack: Vec<PackageId>,
+        resolve: &'a ops::WorkspaceResolve,
+        args: &'a TreeCommandArgs,
+    }
+
+    impl Visitor<'_> {
+        fn visit(
+            &mut self,
+            package_id: PackageId,
+            tree: &mut Tree,
+            depth: usize,
+        ) -> Result<(), BuildError> {
+            self.stack.push(package_id);
+            let result = self.visit_inner(package_id, tree, depth);
+            self.stack.pop();
+            result
+        }
+
+        fn visit_inner(
+            &mut self,
+            package_id: PackageId,
+            tree: &mut Tree,
+            depth: usize,
+        ) -> Result<(), BuildError> {
+            // Only show the core package if explicitly requested.
+            if package_id.is_core() && !self.args.core {
+                return Err(BuildError::Pruned);
+            }
+
+            // Skip if this package should be pruned.
+            if self.args.prune.contains(&package_id.name) {
+                return Err(BuildError::Pruned);
+            }
+
+            // Check if we've reached the maximum depth.
+            // For max_depth=0, we want to list roots only, hence we use strict equality.
+            if let Some(max_depth) = self.args.depth {
+                if depth > max_depth {
+                    tree.max_depth_reached = true;
+                    return Err(BuildError::MaxDepthReached);
+                }
+            }
+
+            tree.package = Some(package_id);
+
+            // Recursively visit dependencies if we haven't visited this package before
+            // or if no_dedupe is true, or if there is a cycle.
```

Comment:
```suggestion
            // or if no_dedupe is true, unless there is a cycle.
```

Nit: I guess? 🤔

Reply:
grammar is hard here, ideally I'd need parentheses, because what I want here is:

```
(haven't visited this package before || if no_dedupe is true) || there is a cycle
```

code has this logic demorganed